zavod/zavod/archive.py

```diff
@@ -24,22 +37,96 @@
 INDEX_FILE = "index.json"
 
 
+class ProtocolBlob(Protocol):
+    def open(self: Blob, mode: str, chunk_size: int) -> None:
+        ...
+
+    def download_to_filename(self, dst: str) -> None:
+        ...
+
+    def reload(self) -> None:
+        ...
+
+
+class Backend(Protocol):
+    def get_blob(self, name: str) -> ProtocolBlob:
+        ...
+
+
+class ConfigurationException(Exception):
+    def __init__(self, message: str) -> None:
+        self.message = message
+
+
+class CloudStorageBackend:
+    def __init__(self) -> None:
+        if settings.ARCHIVE_BUCKET is None:
+            raise ConfigurationException("No backfill bucket configured")
+        client = Client()
+        self.bucket = client.get_bucket(settings.ARCHIVE_BUCKET)
+
+    def get_blob(self, name: str) -> Blob:
+        return self.bucket.get_blob(name)
+
+
+# Other than being a bit icky, there's no real demand for abstracting away
+# the Google Cloud Storage API right now, so this is a blatant mirror for now.
+class FileSystemBlob:
+    def __init__(self, base_path: Path, name: str) -> None:
+        self.path = base_path / name
+        self.name = name
+
+    def open(self, mode: str, chunk_size: int) -> IO[Any]:
+        log.info(f"Opening {self.path}")
+        return open(self.path, mode, buffering=chunk_size)
+
+    def download_to_filename(self, dst: str) -> None:
+        log.info(f"Copying file {self.path} to {dst}")
+        shutil.copyfile(self.path, dst)
+
+    def reload(self) -> None:
+        pass
+
+
+class FilesystemBackend:
+    def __init__(self) -> None:
+        if settings.ARCHIVE_PATH is None:
+            raise ConfigurationException("No archive path configured.")
+        self.base_path = Path(settings.ARCHIVE_PATH)
+
+    def get_blob(self, name: str) -> Optional[FileSystemBlob]:
+        path = self.base_path / name
+        if os.path.isfile(path):
+            return FileSystemBlob(self.base_path, name)
+        else:
+            log.info(f"File {path} doesn't exist.")
+            return None
+
+
+backends = {
+    "CloudStorageBackend": CloudStorageBackend,
+    "FilesystemBackend": FilesystemBackend,
+}
+
+
 @cache
-def get_archive_bucket() -> Optional[Bucket]:
-    if settings.ARCHIVE_BUCKET is None:
-        log.warn("No backfill bucket configured")
+def get_archive_backend() -> Optional[Backend]:
+    if settings.ARCHIVE_BACKEND is None:
+        log.info("No backfill backend configured.")
+        return None
+    try:
+        return cast(Backend, backends[settings.ARCHIVE_BACKEND]())
```

Comment:
I think this cast means we don't actually get type checking between the backends, but at least the protocol provides a bit of documentation. Not sure how to do this better right now. Nominal typing might be nicer in future